content/blog/2020-04-dplyr-1-0-0-and-vctrs/index.Rmarkdown

```diff
@@ -0,0 +1,197 @@
+
+## Base R behaviour
+
+You might wonder why we can't just copy the behaviour of `c()`. Unfortunately `c()` has some major downsides:
+
+*   It doesn't possess a `factor` method so it converts factors to their
+    underlying integer levels. 
+    
+    ```{r}
+    c(factor("x"), factor("y"))
+    ```
+
+*   It's difficult to implement methods when different classes are involved.
+    For example, combining a date (`Date`) and a date-time (`POSIXct`) yields
+    an incorrect result because the underlying data is combined without
+    first being translated.
+
+    ```{r}
+    today <- as.Date("2020-03-24")
+    now <- as.POSIXct("2020-03-24 10:34")
+    
+    c(today, now)
+    # (the second value is the 11 Dec 4341727-12-11)
+    class(c(today, now))
+    unclass(c(today, now))
+    
+    c(now, today)
+    class(c(now, today))
+    unclass(c(now, today))
+    ```
+
+It's difficult to change how `c()` works because any changes are likely to break some existing code, and base R is committed to backward compatibility. Additionally, `c()` isn't the only way that base R combines vectors. `rbind()` and `unlist()` can also be used to perform a similar job, but return different results. This is not to say that the tidyverse has been any better in the past --- we have used a variety of ad hoc methods, undoubtedly using well more than three different approaches. 
+
+Given that it's hard to fix the problem in base R, we've come up with our own alternative to `c()`: `vectrs::vec_c()`. `vec_c()`'s behaviour is governed by three main principles:
+
+*   Symmetry: `vec_c(x, y)` should return a type as similar as possible to 
+    `vec_c(y, x)`. For example, when combining a date and a date-time you
+    always get a date-time.
+    
+    ```{r}
+    vec_c(today, now)
+
+    vec_c(now, today)
+    ```
+    
+*   Enrichment: `vec_c(x, y)` should return the richer type, where type `<x>` 
+    is richer than type `<y>` if `x` can represent all values in `y`. For 
+    example, this implies that combining an integer and double should return a
+    double, and that combining a date and date-time should return a date-time.
+  
+    ```{r}
+    vec_c(1, 1.5)
+    vec_c(today, now)
+    ```
+    
+*   Consistency: `vec_c(x, y)` should error if `x` and `y` are of fundamentally 
+    different types. For example, this implies that combining a string and a
+    number or a factor and a date should error.
+
+    ```{r, error = TRUE}
+    vec_c("a", 1)
+    vec_c(factor("x"), today)
+    ```
+
+## Errors
+
+As a data scientist, you don't really need to much about the vctrs package, except that it exists and its used internally by dplyr. (As a software engineer, you might want to learn about vctrs because [it makes it easier to create new types of vector](https://vctrs.r-lib.org/articles/s3-vector.html)). But vctrs is responsible for creating a number of error messages in dplyr, so it's worth understanding their basic form.
```

Comment:
```suggestion
As a data scientist, you don't really need to much about the vctrs package, except that it exists and its used internally by dplyr. (As a software engineer, you might want to learn about vctrs because [it makes it easier to create new types of vectors](https://vctrs.r-lib.org/articles/s3-vector.html)). But vctrs is responsible for creating a number of error messages in dplyr, so it's worth understanding their basic form.
```